daffodil-lib/src/main/scala/org/apache/daffodil/util/Cursor.scala

```diff
@@ -160,18 +196,24 @@ trait CursorImplMixin[AccessorType <: Accessor[AccessorType]] { self: Cursor[Acc
     val res = priorOpKind match {
       case Advance => {
         priorOpKind = Inspect
-        doAdvance(true)
+        doAdvance(true, true)
       }
-      case Inspect => true // inspect again does nothing.
+      case InspectPure => {
+        priorOpKind = Inspect
+        doAdvance(true, advanceInput=false)
+      }
+      case Inspect      => true // inspect again does nothing.
       case Unsuccessful => return false
     }
     // successful
     Assert.invariant(isFilled == true || !res)
     res
   }
 
-  private def doAdvance(isFilledValue: Boolean) = {
-    val res = fill()
+  //inpsectPure implemented in InfosetInputter.scala
+
+  private def doAdvance(isFilledValue: Boolean, advanceInput: Boolean) = {
```

Comment:
Having a method doAdvance with a flag argument called advanceInput is pretty confusing.  Can you provide a comment to clarify what side effects doAdvance does that aren't suppressed by advanceInput false?

Reply:
Again, we have a really wonky API here. I'll add a comment in this part of the code explaining, but for the sake of discussion I'll include an explanation here to.

Logically speaking, there are two "streams" that we need to keep in our mind when thinking about an InfosetCursor.

There is the input stream, which contains a series of infoset events, and the output stream, which is a sequence of infoset nodes.

The first time you observe an infoset node, you consume it from the input stream. However, it is not consumed from the output stream until the "last" time you observe it (eg. you observe it through calling advance()). The advanceInput flag is used to indicate if the call to advance() is also the first time the node is being observed, in which case it also needs to consume the relevent portion of the input stream.

Arguably, this logic really belongs in the priorOpKind state machine we have sprinkled throughout the advance/inspect/inspectPure methods.

Reply:
The point of cursor is that unlike hasNext() and next() style APIs, one can advance cursors without any little crud objects being allocated. You advance the cursor and an accessor is modified to provide access to what is now "under" the cursor. This is ordinary APIs for big-data, e.g., database cursors. Inspect gives a 1 object look-ahead capability.  What does inspectPure add to this? Does it NOT provide access to the data being inspected? Is it just a test for whether there is or isn't a next?

Reply:
The issue is that in the InfosetCursor implementation, constructing the object has side effects. So, in order to use the 1 object lookahead, the side effects associated with creating an object occur when you use the lookahaed, instead of when you actually consume the object. This means that there is no way to inspect any aspect of the next object without triggering the side effects associated with creating it.

InspectPure provides a limited 1 object look ahead that can be used without triggering the side effects associated with constructing the full object like inspect would.

In more practical terms, inspectPure provides only the ElementRuntimeData, while inspect provides a DINode.

This is actually orthogonal to the difference between more traditional iterators. We would have the same problem with a next()/hasNext()/peek() style API, as our implementation of peek() would have the same side effects as inspect() does now.

The API of Cursor itself is actuall pretty sane. The problem comes when we make InfosetCursor use Accessor[DINode], as that requires making inspect()/next() have unexpected side-effects.